### ISSUE TITLE
cuDNN half-precision batch normalization binding does not work

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#4547]
The VariableType binding code for cuDNN batch norm looks like:

![26695836_146541072795362_1826711580_o](https://user-images.githubusercontent.com/655866/34699474-835d6eea-f4ab-11e7-9374-f5cff4dfa2c1.png)

The problem is that inputs are of mixed types, but `unpack` only handles types that exactly match `input`. We should probably use `unpack_any` here.

As reported by Yuxin

**Comment by ezyang** [comment]
So, how do we spell this in ATen `native_functions.yaml`? Similarly, can we stop calling `ByteTensor` `BoolTensor`?

**Comment by ngimel** [comment]
Huh, it used to work for us before #4285, I guess we should finally add a test ;-)

**Comment by colesbury** [comment]
Longer term we should stop checking precise types in VariableType.cpp, and rely on the underlying `Type` to check precise types (float vs. double). We're not doing this now because `unpack_any` is slower than `unpack`.

Short term, we should just do whatever to make this function work. Special-casing in `gen_variable_type.py` is fine with me.
